api/src/routes/donate.ts

```diff
@@ -10,71 +15,209 @@ import {
-          data: {
-            isDonating: true
+    fastify.post(
+      '/donate/charge-stripe-card',
+      {
+        schema: schemas.chargeStripeCard
+      },
+      async (req, reply) => {
+        try {
+          const { paymentMethodId, amount, duration } = req.body;
+          const { id } = req.session.user;
+
+          const user = await fastify.prisma.user.findUnique({
+            where: { id }
+          });
+
+          if (!user) {
+            void reply.code(500);
+            return {
+              type: 'danger',
+              message: 'Something went wrong.'
+            } as const;
           }
-        });
-
-        return {
-          isDonating: true
-        } as const;
-      } catch (error) {
-        fastify.log.error(error);
-        void reply.code(500);
-        return {
-          type: 'danger',
-          message: 'Something went wrong.'
-        } as const;
+
+          const { email, name } = user;
+
+          if (user.isDonating) {
+            void reply.code(400);
+            return {
+              type: 'info',
+              message: 'User is already donating.'
+            } as const;
+          }
+
+          // Create Stripe Customer
+          let customerId;
+          try {
+            const customer = await stripe.customers.create({
+              email,
+              payment_method: paymentMethodId,
+              invoice_settings: { default_payment_method: paymentMethodId },
+              ...(name && { name })
+            });
+            customerId = customer?.id;
+          } catch {
+            throw {
+              type: 'customerCreationFailed',
+              message: 'Failed to create stripe customer'
+            };
+          }
+
+          // //Create Stripe Subscription
+          let subscriptionId;
+
+          const plan = `${donationSubscriptionConfig.duration[
+            duration
+          ].toLowerCase()}-donation-${amount}`;
+
+          try {
+            const {
+              id: subscription_id,
+              latest_invoice: {
+                // following key is not present in new api definitions. @ts-ignore as recommended by Stripe docs
+                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+                // @ts-ignore stripe-version-2019-10-17
+                payment_intent: { client_secret, status }
+              }
+            } = await stripe.subscriptions.create({
+              customer: customerId,
+              payment_behavior: 'allow_incomplete',
+              items: [{ plan }],
+              expand: ['latest_invoice.payment_intent']
+            });
+            if (status === 'requires_source_action') {
+              void reply.code(402);
+              return {
+                type: 'UserActionRequired',
+                message: 'Payment requires user action',
+                // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+                client_secret
```

Comment:
Where is this variable coming from?

Also, if this is our Stripe secret, are we sure we want to return this in an API response?

Reply:
If a card need authentication, we receive an 'requires_source_action' error. 
In that case we take the client_secret that is related to that payment intent and pass it to the related function on client to process the card : https://github.com/freeCodeCamp/freeCodeCamp/blob/c0f732063f2d0273adc703833ab60274e09b914a/client/src/components/Donation/stripe-card-form.tsx#L130